Install lxc from source.

jessfraz opened the pull request:
Haters gunna hate, cache bust. But this fixes a bunch of tests.

Specifically the volumes tests of #9875:

```
- TestRunVolumesMountedAsReadonly (0.95s)
- TestRunVolumesFromInReadonlyMode (2.08s)
- TestRunVolumesFromInReadWriteMode (0.03s)
- TestVolumesFromGetsProperMode (0.02s)
- TestRunApplyVolumesFromBeforeVolumes (0.02s)
```

Improvement! Yayyyy!
After we get the CI, running in success we will know when we break things!

Also this way we can stay up with the latest lxc and hopefully there will be fixes for the other tests :)

ping @ashahab-altiscale

jessfraz committed:
Install lxc from source.

Haters gunna hate, cache bust. But this fixes a bunch of tests.

Docker-DCO-1.1-Signed-off-by: Jessica Frazelle <jess@docker.com> (github: jfrazelle)

tianon commented:
So...  Does this mean we need to raise the minimum bar again in
PACKAGERS.md?

ashahab-altiscale commented:
@jfrazelle This looks good! Thanks. 
@tianon Yes, I think we should recommend installing lxc-1.0.7.

jfrazelle commented:
I can add that to this PR, do it all in one swoop

On Wednesday, January 7, 2015, Abin Shahab <notifications@github.com> wrote:

> @jfrazelle <https://github.com/jfrazelle> This looks good! Thanks.
> @tianon <https://github.com/tianon> Yes, I think we should recommend
> installing lxc-1.0.7.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9947#issuecomment-69050521>.
>

jfrazelle commented:
I don't know if we wanna write a little bit more in PACKAGERS, I'm open to it, I'm just not really sure what to say

jessfraz committed:
Update the LXC version to 1.0.7 in PACKAGERS.md

Signed-off-by: Jessica Frazelle <jess@docker.com>

Docker-DCO-1.1-Signed-off-by: Jessica Frazelle <jess@docker.com> (github: jfrazelle)

ashahab-altiscale commented on project/PACKAGERS.md:
```diff
@@ -306,7 +306,7 @@ the client will even run on alternative platforms such as Mac OS X / Darwin.
 by having support for them in the kernel or userspace. A few examples include:
 
-* LXC execution driver (requires version 1.0 or later of the LXC utility scripts)
+* LXC execution driver (requires version 1.0.7 or later of the LXC utility scripts)
```
Not "LXC utility scripts". IMHO, docker specifically needs lxc and lxc-libs

jfrazelle replied:
ah ok cool will update

On Wed, Jan 7, 2015 at 9:48 AM, Abin Shahab <notifications@github.com>
wrote:

> In project/PACKAGERS.md
> <https://github.com/docker/docker/pull/9947#discussion-diff-22601688>:
>
> > @@ -306,7 +306,7 @@ the client will even run on alternative platforms such as Mac OS X / Darwin.
> >  Some of Docker's features are activated by using optional command-line flags or
> >  by having support for them in the kernel or userspace. A few examples include:
> >
> > -* LXC execution driver (requires version 1.0 or later of the LXC utility scripts)
> > +* LXC execution driver (requires version 1.0.7 or later of the LXC utility scripts)
>
> Not "LXC utility scripts". IMHO, docker specifically needs lxc and lxc-libs
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9947/files#r22601688>.
>

jfrazelle commented:
unsure if it's supposed to be all caps or not...

crosbymichael commented:
LGTM

SvenDowideit commented:
LGTM, not that i have a say :)

LK4D4 commented:
LGTM

LK4D4 commented:
@tianon It is for you to merge :) So we can blame you later.

tianon commented:
We'll all blame me regardless. :)

Have we tested on 1.0.7 to verify that all the tests actually pass at that
low of a version?  If they don't, I think our minimum will need to be
bumped up further. ;)

tianon commented:
That's what I get for commenting via email instead of reading the PR diff...

tianon commented on Dockerfile:
```diff
@@ -62,6 +61,13 @@ RUN cd /usr/local/lvm2 \
+# Install lxc
+RUN curl -sSL https://linuxcontainers.org/downloads/lxc/lxc-1.0.7.tar.gz | tar -v -C /usr/src/ -xz
+RUN cd /usr/src/lxc-1.0.7 \
+    && ./configure --prefix=/usr/ \
```
These are space-indented. :wink:

tianon commented on Dockerfile:
```diff
@@ -62,6 +61,13 @@ RUN cd /usr/local/lvm2 \
 # see https://git.fedorahosted.org/cgit/lvm2.git/tree/INSTALL
 
+# Install lxc
+RUN curl -sSL https://linuxcontainers.org/downloads/lxc/lxc-1.0.7.tar.gz | tar -v -C /usr/src/ -xz
```
Wouldn't this be slightly easier to change versions if we used `tar`'s `--strip-components` so the version number is only hard coded in one spot?

jfrazelle replied:
/me reads tar man pages,
brain explodes where the eff has this been all my life

tianon commented on Dockerfile:
```diff
@@ -62,6 +61,13 @@ RUN cd /usr/local/lvm2 \
+RUN cd /usr/src/lxc-1.0.7 \
+    && ./configure --prefix=/usr/ \
+    && make \
+    && sudo make install
```
What the sudo?

jfrazelle replied:
blerg copy pasta lol

On Wed, Jan 7, 2015 at 10:01 PM, Tianon Gravi <notifications@github.com>
wrote:

> In Dockerfile
> <https://github.com/docker/docker/pull/9947#discussion-diff-22636754>:
>
> > @@ -62,6 +61,13 @@ RUN cd /usr/local/lvm2 \
> >  	&& make install_device-mapper
> >  # see https://git.fedorahosted.org/cgit/lvm2.git/tree/INSTALL
> >
> > +# Install lxc
> > +RUN curl -sSL https://linuxcontainers.org/downloads/lxc/lxc-1.0.7.tar.gz | tar -v -C /usr/src/ -xz
> > +RUN cd /usr/src/lxc-1.0.7 \
> > +    && ./configure --prefix=/usr/ \
> > +    && make \
> > +    && sudo make install
>
> What the sudo?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9947/files#r22636754>.
>

jfrazelle commented:
ok updated

jfrazelle commented on Dockerfile:
```diff
@@ -62,6 +61,14 @@ RUN cd /usr/local/lvm2 \
 # see https://git.fedorahosted.org/cgit/lvm2.git/tree/INSTALL
 
+# Install lxc
+RUN mkdir -p /usr/src/lxc
```
I kinda hate this now tho, should we put it in one run w the curl

crosbymichael replied:
Layers are evil

tianon replied:
Yeah, I agree, one layer seems simplest and cleanest here.

tianon commented on Dockerfile:
```diff
@@ -62,6 +61,14 @@ RUN cd /usr/local/lvm2 \
+RUN mkdir -p /usr/src/lxc
+RUN curl -sSL https://linuxcontainers.org/downloads/lxc/lxc-1.0.7.tar.gz | tar -v -C /usr/src/lxc/ -xz --strip-components=1
+RUN cd /usr/src/lxc \
+	&& ./configure --prefix=/usr/ \
```
Shouldn't this be `--prefix=/usr/local`?

jfrazelle replied:
no it fails to use lxc correctly in the container if I do that

jfrazelle replied:
updated!

On Thu, Jan 8, 2015 at 10:49 AM, Tianon Gravi <notifications@github.com>
wrote:

> In Dockerfile
> <https://github.com/docker/docker/pull/9947#discussion-diff-22671703>:
>
> > @@ -62,6 +61,14 @@ RUN cd /usr/local/lvm2 \
> >  	&& make install_device-mapper
> >  # see https://git.fedorahosted.org/cgit/lvm2.git/tree/INSTALL
> >
> > +# Install lxc
> > +RUN mkdir -p /usr/src/lxc
> > +RUN curl -sSL https://linuxcontainers.org/downloads/lxc/lxc-1.0.7.tar.gz | tar -v -C /usr/src/lxc/ -xz --strip-components=1
> > +RUN cd /usr/src/lxc \
> > +	&& ./configure --prefix=/usr/ \
>
> Shouldn't this be --prefix=/usr/local?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9947/files#r22671703>.
>

tianon replied:
That sounds like a bug! D:

jfrazelle replied:
well their default for .configure is /usr/local as the prefix and I couldn't get it to work with that.... idk

tianon replied:
D:

I'm cloning now to take a look and see if I can't help.

tianon commented:
```patch
diff --git a/Dockerfile b/Dockerfile
index b869594..b38c142 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -65,9 +65,10 @@ RUN cd /usr/local/lvm2 \
 RUN mkdir -p /usr/src/lxc \
        && curl -sSL https://linuxcontainers.org/downloads/lxc/lxc-1.0.7.tar.gz | tar -v -C /usr/src/lxc/ -xz --strip-components=1
 RUN cd /usr/src/lxc \
-       && ./configure --prefix=/usr/ \
+       && ./configure \
        && make \
-       && make install
+       && make install \
+       && ldconfig
 
 # Install Go
 RUN curl -sSL https://golang.org/dl/go1.4.src.tar.gz | tar -v -C /usr/local -xz
```
:tada:

jfrazelle commented:
niceeeeeee

jfrazelle commented:
updated!

jfrazelle commented:
wdwt cache busting time :D

tianon commented:
LGTM! Bust it!

jfrazelle commented:
![burn](https://hugthief.s3.amazonaws.com/gifs/burnitdown.gif)

ashahab-altiscale commented:
Thanks a lot for this!
On Jan 8, 2015 5:39 PM, "Jessie Frazelle" <notifications@github.com> wrote:

> [image: burn]
> <https://camo.githubusercontent.com/a8d5775a627d551bf1aecb8002b5b6eb1247b56a/68747470733a2f2f68756774686965662e73332e616d617a6f6e6177732e636f6d2f676966732f6275726e6974646f776e2e676966>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9947#issuecomment-69281407>.
>
